### ISSUE TITLE
Return Text Fragments for linking to highlights?

### ISSUE THREAD

**Issue by karlhorky** [tricinel/highlight-words#33]
Hey @tricinel 👋 hope things are good with you!

I wanted to see whether you'd be open to also returning a [Text Fragment](https://wicg.github.io/scroll-to-text-fragment/) ([blog post](https://web.dev/articles/text-fragments)) in addition to the chunks matched in a string.

Eg. 

1. Returning the Text Fragment `Modules in Web` for usage in the URL below, for matching and scrolling to that part of the page:
   https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#:~:text=Modules%20in%20Web
   ![Screenshot 2023-10-15 at 22 29 35](https://github.com/bvaughn/highlight-words-core/assets/1935696/ce802fd0-f165-4427-b009-d0c94420ed5e)<br /><br />
2. Returning the Text Fragment `details%20about%20the-,API,-%2C%20see%20Experimenting%20with` for usage in the URL below, for matching a specific instance of API in the page:
   https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#:~:text=details%20about%20the-,API,-%2C%20see%20Experimenting%20with
   ![Screenshot 2023-10-15 at 22 32 50](https://github.com/bvaughn/highlight-words-core/assets/1935696/94958be1-a393-47fb-9551-ec485aa95c96)

Ideas for implementation can be inspired by the `doGenerateFragment` function in `fragment-generation-utils.js` from the [`text-fragments-polyfill`](https://www.npmjs.com/package/text-fragments-polyfill) package:

https://github.com/GoogleChromeLabs/text-fragments-polyfill//blob/e5252cb6eba768cdc4166b27cbe8080fc67f37c7/src/fragment-generation-utils.js#L145-L243

**Comment by tricinel** [comment]
Probably not at this time. What would be the use case? Could you give me an example?

**Comment by karlhorky** [comment]
It would be a UX feature of search:

- Backend searches through content of multiple pages using `highlight-words`
- The `textFragment` in the `SearchResult` is used to link to the unique location in one of those pages, scrolling down the page

**Comment by tricinel** [comment]
So on the back end, you're searching some content and getting back a list of matches/non-matches, wrapped in a link with the hash and the text fragment.

So why can't you use it as is now? You're already getting back a list of matches/non-matches. So you could forEach match and wrap it however you want. What am I missing?

**Comment by karlhorky** [comment]
> getting back a list of matches/non-matches, wrapped in a link with the hash and the text fragment

Yes, to be clear, it's not the list itself that is wrapped in the link, it's each item in the list.

> What am I missing?

Not entirely sure where the gaps are but I'll try to re-explain in more detail.

The Text Fragment (the thing after the `#:~:text=`, as mentioned in the issue description above) is not yet generated - it needs to be generated somewhere. Generating a Text Fragment which is unique across the whole document can be done with code like [this](https://github.com/GoogleChromeLabs/text-fragments-polyfill//blob/e5252cb6eba768cdc4166b27cbe8080fc67f37c7/src/fragment-generation-utils.js?rgh-link-date=2023-10-15T20%3A35%3A59Z#L145-L243), but this code deals with selections on the page, which are not relevant for the search UX.

Given that `highlight-words` has the whole document already in memory and is already splitting it into chunks, it is potentially in a good position to create a unique Text Fragment so that a link can be created to this specific highlight.

As of now, the data passed back in the chunks offers a rudimentary way of creating a link (`${previousChunk}-,${currentChunk},-${nextChunk}`), but this naive approach is not robust - one first case where they fail is when a phrase is repeated in the document.

**Comment by tricinel** [comment]
Understood. I don't think personally this helps others in many ways to be honest. The argument could be made, if you already have the document, why not return some JSX, or a vue component, or a lit element. But that's not the purpose of this. We just split the content into chunks that are either a match or not a match. And the consumer can manipulate it afterwards as they please.

In my own use cases, I either count the matches (so I don't need extra work happening inside of the matcher for some transformation I won't use) or I apply some CSS on chunks conditionally (again, that work should not happen in the matcher).

I understand your use case. The uniqueness of the fragment will be problematic either in the library or outside of it, so I don't think that's an argument either - if the phrase is repeated in the document, it will be the same problem in the matching function as it will in your app.

Bottom line, I don't see a point for this right now. We can revisit later if need be.
